src/fastify-plugins/maps/static-maps.js

```diff
@@ -0,0 +1,284 @@
+      rep.header('Content-Encoding', 'gzip')
+    }
+
+    rep.send(data)
+  }
+
+  // Serve static files
+  fastify.register(FastifyStatic, {
+    root: staticRootDir,
+    setHeaders: (res, path) => {
+      if (path.toLowerCase().endsWith('.pbf')) {
+        res.setHeader('Content-Type', 'application/x-protobuf')
+      }
+    },
+  })
+
+  /// List static maps
+  fastify.get(
+    '/',
+    { schema: { response: { 200: ListStaticMapsReplySchema } } },
+    async (req) => {
+      const styleDirFiles = await fs.readdir(staticRootDir)
+
+      const result = (
+        await Promise.all(
+          styleDirFiles.map(async (filename) => {
+            const stat = await fs.stat(path.join(staticRootDir, filename))
+            if (!stat.isDirectory()) return null
+
+            let styleJson
+
+            try {
+              const styleJsonContent = await fs.readFile(
+                path.join(staticRootDir, filename, 'style.json'),
+                'utf-8'
+              )
+
+              styleJson = JSON.parse(styleJsonContent)
+            } catch (err) {
+              return null
+            }
+
+            const serverAddress = await getFastifyServerAddress(
+              req.server.server
+            )
+
+            return {
+              id: filename,
+              name: typeof styleJson.name === 'string' ? styleJson.name : null,
+              styleUrl: new URL(
+                `${req.server.prefix || ''}/${filename}/style.json`,
+                serverAddress
+              ).href,
+            }
+          })
+        )
+      ).filter(
+        /**
+         * @template {import('@sinclair/typebox').Static<typeof ListStaticMapsReplySchema>[number] | null} V
+         * @param {V} v
+         * @returns {v is NonNullable<V>}
+         */
+        (v) => v !== null
+      )
+
+      return result
+    }
+  )
+
+  /// Get a map's style.json
+  fastify.get(
+    `/:styleId/style.json`,
+    { schema: { params: GetStyleJsonParamsSchema } },
+    async (req, rep) => {
+      const { styleId } = req.params
+
+      const serverAddress = await getFastifyServerAddress(req.server.server)
+
+      let stats, styleJson
+
+      try {
+        const results = await Promise.all([
+          fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
+          fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
+        ])
+
+        stats = results[0]
+        styleJson = results[1]
+      } catch (err) {
+        throw new NotFoundError(`id = ${styleId}, style.json`)
+      }
+
+      rep.headers({
+        'Content-Type': 'application/json; charset=utf-8',
+        'Cache-Control': 'max-age=' + 5 * 60, // 5 minutes
+        'Access-Control-Allow-Headers':
+          'Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since',
+        'Access-Control-Allow-Origin': '*',
+        'Last-Modified': new Date(stats.mtime).toUTCString(),
+        'Content-Length': Buffer.from(styleJson).length,
```

Comment:
tl;dr: this could be improved slightly but is fine. Feel free to ignore the rest of this comment.

---

Nitpick: there are some unnecessary byte-string conversions in this flow:

1. Load *bytes* from disk, convert them to a *string* (using `fs.readFile(..., 'utf-8')`).
1. Convert the *string* to *bytes* to compute the `Content-Length` (this line).
1. Convert the *string* to *bytes* to send the data.

In practice, these conversions probably have minimal impact, and it's unlikely that the file will ever be too long to fit in a string[^0]. But in my deranged mind, this tiny imperfection cannot stand.

<details>
<summary>I implemented a full solution that does all operations on Buffers:</summary>

```diff
diff --git a/src/fastify-plugins/maps/static-maps.js b/src/fastify-plugins/maps/static-maps.js
index e6927ee..96fe965 100644
--- a/src/fastify-plugins/maps/static-maps.js
+++ b/src/fastify-plugins/maps/static-maps.js
@@ -7,6 +7,7 @@ import asar from '@electron/asar'
 import { Mime } from 'mime/lite'
 import standardTypes from 'mime/types/standard.js'
 
+import { bufferReplace } from '../../utils.js'
 import { NotFoundError, getFastifyServerAddress } from '../utils.js'
 
 export const PLUGIN_NAME = 'mapeo-static-maps'
@@ -24,7 +25,7 @@ export const plugin = fp(staticMapsPlugin, {
 
 /**
  * @typedef {object} StaticMapsPluginDecorator
- * @property {(styleId: string, serverAddress: string) => Promise<string>} getResolvedStyleJson
+ * @property {(styleId: string, serverAddress: string) => Promise<Buffer>} getResolvedStyleJson
  * @property {(styleId: string) => Promise<import('node:fs').Stats>} getStyleJsonStats
  */
 
@@ -34,10 +35,11 @@ async function staticMapsPlugin(fastify, opts) {
     async getResolvedStyleJson(styleId, serverAddress) {
       const filePath = path.join(opts.staticRootDir, styleId, 'style.json')
 
-      const data = await fs.readFile(filePath, 'utf-8')
+      const data = await fs.readFile(filePath)
 
-      return data.replace(
-        /\{host\}/gm,
+      return bufferReplace(
+        data,
+        '{host}',
         new URL(`${opts.prefix || ''}/${styleId}`, serverAddress).href
       )
     },
@@ -199,7 +201,7 @@ async function routes(fastify, opts) {
           'Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since',
         'Access-Control-Allow-Origin': '*',
         'Last-Modified': new Date(stats.mtime).toUTCString(),
-        'Content-Length': Buffer.from(styleJson).length,
+        'Content-Length': styleJson.byteLength,
       })
 
       return styleJson
diff --git a/src/utils.js b/src/utils.js
index 8ba6d0e..69d2ce5 100644
--- a/src/utils.js
+++ b/src/utils.js
@@ -123,6 +123,52 @@ export function deNullify(obj) {
   return /** @type {import('./types.js').NullableToOptional<T>} */ (objNoNulls)
 }
 
+/**
+ * @param {Readonly<string | Buffer>} stringOrBuffer
+ * @returns {Buffer}
+ */
+function bufferify(stringOrBuffer) {
+  return Buffer.isBuffer(stringOrBuffer)
+    ? stringOrBuffer
+    : Buffer.from(stringOrBuffer)
+}
+
+/**
+ * @param {Readonly<Buffer>} buffer
+ * @param {Readonly<string | Buffer>} pattern
+ * @param {Readonly<string | Buffer>} replacement
+ * @returns {Buffer}
+ */
+export function bufferReplace(buffer, pattern, replacement) {
+  const patternLength = bufferify(pattern).byteLength
+  replacement = bufferify(replacement)
+
+  /** @type {Buffer[]} */
+  const chunks = []
+  let totalLength = 0
+
+  /** @param {Buffer} chunk */
+  const addChunk = (chunk) => {
+    chunks.push(chunk)
+    totalLength += chunk.byteLength
+  }
+
+  while (buffer.byteLength) {
+    const index = buffer.indexOf(pattern)
+    if (index === -1) {
+      addChunk(buffer)
+      break
+    }
+
+    addChunk(buffer.subarray(0, index))
+    addChunk(replacement)
+
+    buffer = buffer.subarray(index + patternLength)
+  }
+
+  return Buffer.concat(chunks, totalLength)
+}
+
 /**
  * @template {import('@mapeo/schema').MapeoDoc & { forks?: string[] }} T
  * @param {T} doc
diff --git a/tests/utils.js b/tests/utils.js
index e80a195..a3e4871 100644
--- a/tests/utils.js
+++ b/tests/utils.js
@@ -1,6 +1,11 @@
 // @ts-check
 import test from 'brittle'
-import { assert, ExhaustivenessError, setHas } from '../src/utils.js'
+import {
+  assert,
+  ExhaustivenessError,
+  setHas,
+  bufferReplace,
+} from '../src/utils.js'
 
 test('assert()', (t) => {
   t.execution(() => assert(true, 'should work'))
@@ -27,3 +32,16 @@ test('setHas()', (t) => {
   t.ok(setHas(set)(1))
   t.absent(setHas(set)(9))
 })
+
+test('bufferReplace()', (t) => {
+  const fooBarBaz = Buffer.from('foo bar baz')
+
+  const expected = Buffer.from('foo zor zoz')
+  t.alike(bufferReplace(fooBarBaz, 'ba', 'zo'), expected)
+  t.alike(
+    bufferReplace(fooBarBaz, Buffer.from('ba'), Buffer.from('zo')),
+    expected
+  )
+
+  t.alike(bufferReplace(fooBarBaz, 'x', 'ignored'), fooBarBaz)
+})
```
</details>

That's a fairly heavyweight solution, though. Instead, would you move the `Buffer.from(styleJson)` call up so that it can be used to compute the `Content-Length` _and_ for the result? This should remove a needless conversion.

```diff
diff --git a/src/fastify-plugins/maps/static-maps.js b/src/fastify-plugins/maps/static-maps.js
index e6927ee..1b97131 100644
--- a/src/fastify-plugins/maps/static-maps.js
+++ b/src/fastify-plugins/maps/static-maps.js
@@ -192,6 +192,8 @@ async function routes(fastify, opts) {
         throw new NotFoundError(`id = ${styleId}, style.json`)
       }
 
+      const styleJsonBytes = Buffer.from(styleJson)
+
       rep.headers({
         'Content-Type': 'application/json; charset=utf-8',
         'Cache-Control': 'max-age=' + 5 * 60, // 5 minutes
@@ -199,10 +201,10 @@ async function routes(fastify, opts) {
           'Authorization, Content-Type, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since',
         'Access-Control-Allow-Origin': '*',
         'Last-Modified': new Date(stats.mtime).toUTCString(),
-        'Content-Length': Buffer.from(styleJson).length,
+        'Content-Length': styleJsonBytes.length,
       })
 
-      return styleJson
+      return styleJsonBytes
     }
   )
```

[^0]: In my cursory testing, the max length of a string is 536,870,888 characters. Not sure what a "character" means in this case.

Reply:
looks like i simplify the implementation here in this PR: https://github.com/digidem/mapeo-core-next/pull/450/files#diff-38bb092a0481e2a5e86a9d4af8f6fc9fb9c5c1e7283570ef0d146c97be8836a4

so potentially moot to a certain extent 😄